### ISSUE TITLE
I can't migrate PostGIS tables..

### ISSUE THREAD

**Issue by jon890** [prisma/prisma#13127]
### Bug description

I think that verifying with shadow database failed because migrate tools drop table spatial_ref_sys
so public.geometry does not exist..

### How to reproduce

1.make schema with geo table
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider          = "postgresql"
  url               = env("DATABASE_URL")
}

model Geom {
  geomId    Int                                                    @id @default(autoincrement())
  geomData  Unsupported("\"public\".geometry(multipolygon, 4326)")
  createdAt DateTime                                               @default(now()) @db.Timestamptz
  updatedAt DateTime                                               @updatedAt @db.Timestamptz
}

2. npx prisma migrate --name init

3. modify schema
schema.prisma
...
model test {
  id  Int @id @default(autoincrement())
}

4. again npx prisma migrate --name add_test_table

### Expected behavior

working well!

### Prisma information

Error: P3006

Migration `20220503113444_add_geom` failed to apply cleanly to the shadow database. 
Error:
db error: ERROR: type "public.geometry" does not exist
   0: sql_migration_connector::validate_migrations
             at migration-engine/connectors/sql-migration-connector/src/lib.rs:272
   1: migration_core::state::DevDiagnostic
             at migration-engine/core/src/state.rs:250

### Environment & setup

- OS: Mac Os
- Database: Postgresql with PostGIS
- Node.js version: 16

### Prisma Version

```
prisma                  : 3.13.0
@prisma/client          : 3.13.0
Current platform        : darwin
Query Engine (Node-API) : libquery-engine efdf9b1183dddfd4258cd181a72125755215ab7b (at node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli efdf9b1183dddfd4258cd181a72125755215ab7b (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine    : introspection-core efdf9b1183dddfd4258cd181a72125755215ab7b (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary           : prisma-fmt efdf9b1183dddfd4258cd181a72125755215ab7b (at node_modules/@prisma/engines/prisma-fmt-darwin)
Default Engines Hash    : efdf9b1183dddfd4258cd181a72125755215ab7b
Studio                  : 0.459.0
```

**Comment by janpio** [comment]
Did you add some logic to an earlier migration to enable the postgis extension explicitly?
(I am correct that this error message could also be caused by the extension not being loaded, right?)

**Comment by jon890** [comment]
> 

yes we should create extension to enable postgis
I think that shadow database has not been enable postgis, so encounter above errors.

**Comment by janpio** [comment]
Yes. Did you add these commands to a migration so they are part of your migration history? The shadow database is a newly created one, and then only gets the migration files applied. To make sure the shadow database also has the extension enabled, adding the commands to a migration is the right way.

**Comment by sauloaguiar** [comment]
@janpio suggestion worked for me!

**Comment by gbpaixao** [comment]
@janpio I've been having the same issue here, tried the proposed solution and it kinda worked. But whenever I try to add new migrations, I get the same error. 

Here's the error itself when I run prisma migrate dev (--create-only or not):
```
Migration `20220517194741_init` failed to apply cleanly to the shadow database. 
Error:
db error: ERROR: type "public.hstore" does not exist
   0: migration_core::state::DevDiagnostic
             at migration-engine/core/src/state.rs:250
```

Here's the migration called init:

```
-- CreateExtension
CREATE EXTENSION IF NOT EXISTS hstore;

-- CreateExtension
CREATE EXTENSION IF NOT EXISTS unaccent;

...

-- CreateTable
CREATE TABLE "logged_actions" (
    "id" BIGSERIAL NOT NULL,
    "row_data" "public".hstore,

    CONSTRAINT "logged_actions_pkey" PRIMARY KEY ("id")
);

```

Also, here's my schema.prisma:

model logged_actions {
  id                 BigInt                 @id @default(autoincrement())
  row_data           Unsupported("\"public\".hstore")?
}

Any ideia on how to solve this? Thanks in advance.

**Comment by janpio** [comment]
Can you maybe open a quick new issue with just your reproduction? That would make it easier for us to reproduce and discuss. Thanks.

**Comment by humblecoder** [comment]
I'm having a similar issue.  I have a Docker configuration with Postgres in a separate container from the application.  I receive `could not open extension control file` when attempting to enable extensions.

Does the existing capability assume everything is running on the same machine/instance?  

Any chance we could get some sort of schema feature, or `db push` that allows for raw queries (e.g. `CREATE EXTENSION IF NOT EXISTS...`) before any actual table generation?  🤔

**Comment by SevInf** [comment]
Hi everyone!
Can you try it with [`postgresqlExtensions`](https://www.prisma.io/docs/concepts/components/prisma-schema/postgresql-extensions) preview feature? That should fix this issue.

**Comment by nsilvestri** [comment]
I was having the same problem when trying to introduce Prisma to an existing project, where PostGIS was installed on the `public` schema but the application was in a different schema. `prisma migrate reset` would work correctly and instantiate the baseline database with the correct PostGIS types, but I would get the same error `type "public.geometry" does not exist` when using `prisma migrate dev`.

In my `schema.prisma` file, I always refer to all `Unsupported` types with PostGIS's installation schema (in my case, `public`) explicitly declared, like `center        Unsupported("public.geometry(Point, 4326)")`. In my `0-init` migration, I had `CREATE EXTENSION IF NOT EXISTS "postgis";` at the very top, making sure that the extension was explicitly enabled.

To fix the problem, I enabled `postgresqlExtensions`, added `extensions = [postgis(schema: "public")]`, and then changed the line in my migration where I enabled the `postgis` extension to `CREATE EXTENSION IF NOT EXISTS "postgis" WITH SCHEMA "public";`. Now `prisma migrate dev` seems to works as intended.

My incomplete understanding of the shadow database leads me to believe that the problem was that the PostGIS extension in my shadow database was being installed to the application schema (which is specified by querystring parameter in the `DATABASE_URL`), while PostIGS was already installed to `public` in the default database of the `postgis` docker image. `WITH SCHEMA "public";` forced the extension to be installed in the expected schema in the shadow database, and then `public.geometry` could appropriately resolve.

**Comment by lh0x00** [comment]
I’ve created a PR for this feature. I mentioned it before, but I only just found the time to work on it today. If you’re happy with it, feel free to leave a reaction and follow me—really appreciate it!
PRs: https://github.com/prisma/prisma/pull/29365 https://github.com/prisma/prisma-engines/pull/5797 https://github.com/prisma/web/pull/7685

**Comment by thecoolestcoderrushil** [comment]
postgis tables that fail to migrate with p3006 because the shadow database drops spatial_ref_sys are a painful extension bug, since the shadow db does not have postgis enabled and the migration cannot apply cleanly. rehearsing the migration on a production-shaped twin first would surface the extension problem before deploy day. antifailure.dev (https://antifailure.dev) builds disposable database twins for migration rehearsal and is open source and free to try, with demos available for teams that want one.
